core-api/src/main/java/com/optimizely/ab/event/BatchEventProcessor.java

```diff
@@ -122,26 +123,46 @@ public void flush() throws InterruptedException {
         eventQueue.put(FLUSH_SIGNAL);
     }
 
+    private interface QueueService {
+        Object get() throws InterruptedException;
+    }
+
     public class EventConsumer implements Runnable {
         private LinkedList<UserEvent> currentBatch = new LinkedList<>();
         private long deadline = System.currentTimeMillis() + flushInterval;
 
         @Override
         public void run() {
             try {
+                int waitCount = 0;
+
+                QueueService polling = () -> eventQueue.poll(flushInterval, TimeUnit.MILLISECONDS);
+                QueueService take = () -> eventQueue.take();
+                QueueService using = polling;
+
                 while (true) {
                     if (System.currentTimeMillis() > deadline) {
                         logger.debug("Deadline exceeded flushing current batch.");
                         flush();
+                        deadline = System.currentTimeMillis() + flushInterval;
                     }
 
-                    Object item = eventQueue.poll(50, TimeUnit.MILLISECONDS);
+                    Object item = using.get();
```

Comment:
I think the `QueueService` adds a bit too much cognitive load  then what we need in this already complex loop. Since we're basing our decision to use `take()` or `poll(...)` based on the number of times we're exited the poll with `null` items we can use a ternary operator inline:
```java
long timeout = deadline - System.currentTimeMillis();
Object item = emptyCount > MAX_EMPTY_COUNT ? eventQueue.take() : eventQueue.poll(timeout, TimeUnit.MILLISECONDS);
```

Note I'm suggesting renaming from "wait" to "empty" since I think better captures the intention of the counter. I also used the dynamic timeout as opposed to the fix `flushInterval` as we discussed earlier.